cli/src/doctest.rs

```diff
@@ -0,0 +1,417 @@
+
+        // Throw away trailing empty lines.
+        let last_non_empty = lines
+            .iter()
+            .rposition(|line| !line.trim().is_empty())
+            .unwrap_or(0);
+        lines.truncate(last_non_empty + 1);
+
+        let mut expected = Vec::new();
+        for line in lines.iter().rev() {
+            // If we encounter an uncommented line before we find a "=>", there's no expected value
+            // for this test.
+            let Some(commented) = line.trim_start().strip_prefix('#') else {
+                break;
+            };
+
+            if let Some(arrowed) = commented.trim_start().strip_prefix("=>") {
+                // We've found an expected value for the test.
+                if let Some(msg) = arrowed.trim_start().strip_prefix("error:") {
+                    expected.push(msg.trim());
+                    expected.reverse();
+                    return Expected::Error(expected.join("\n"));
+                } else if arrowed.trim() == "error" {
+                    return Expected::Error(String::new());
+                } else {
+                    expected.push(arrowed);
+                    expected.reverse();
+                    return Expected::Value(expected.join("\n"));
+                }
+            } else {
+                expected.push(commented);
+            }
+        }
+
+        Expected::None
+    }
+
+    fn error(&self) -> Option<String> {
+        match self {
+            Expected::Error(s) => Some(s.clone()),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct DocTest {
+    input: String,
+    expected: Expected,
+}
+
+struct TestEntry {
+    expected_error: Option<String>,
+    field_name: LocIdent,
+    test_idx: usize,
+}
+
+#[derive(Default)]
+struct TestRegistry {
+    tests: HashMap<Ident, TestEntry>,
+}
+
+impl DocTest {
+    fn new(input: String) -> Self {
+        let expected = Expected::extract(&input);
+        DocTest { input, expected }
+    }
+
+    fn code(&self) -> String {
+        match &self.expected {
+            Expected::Value(v) => format!("(({}\n)| std.contract.Equal ({v}))", self.input),
+            _ => self.input.clone(),
+        }
+    }
+}
+
+struct Error {
+    path: Vec<LocIdent>,
+    idx: usize,
+    kind: ErrorKind,
+}
+
+enum ErrorKind {
+    UnexpectedFailure {
+        error: EvalError,
+    },
+    /// A doctest was expected to fail, but instead it succeeded.
+    UnexpectedSuccess {
+        result: RichTerm,
+    },
+    /// A doctest failed with an unexpected message.
+    WrongTestFailure {
+        messages: Vec<String>,
+        expected: String,
+    },
+}
+
+// Go through the record spine, running tests one-by-one.
+//
+// `spine` is the already-evaluated record spine, which we use only for finding the path to
+// each test. We don't use `spine` for evaluating the test because `RecRecord`s have already
+// been evaluated to `Record`s so it's challenging to reconstruct the environment.
```

Comment:
I'm not sure to understand this comment. In the end you do use spine, right? Do you mean that you don't use it to evaluate each test invidually, but yo use `eval_closure` instead? But we still provide an empty environment there, so wouldn't spine be fine as well in the end?

Reply:
Hm, I think it was a left-over from an earlier version, before I figured out that closurizing the terms in `eval_record_spine` made everything easier.